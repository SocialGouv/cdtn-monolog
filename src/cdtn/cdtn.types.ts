import { SourceRoute } from "@socialgouv/cdtn-sources";

type Article = {
  _score: number;
  cdtnId: string;
  description: string;
  slug: string;
  source: string;
  title: string;
  url: string;
};

type Document = {
  _score: number;
  algo: string;
  breadcrumbs: { label: string; slug: string }[];
  cdtnId: string;
  description: string;
  slug: string;
  source: SourceRoute;
  title: string;
  url: string;
};

type Theme = {
  _score: number;
  algo: string;
  breadcrumbs?: { label: string; slug: string }[];
  slug: string;
  source: string;
  title: string;
};

type SearchResult = {
  articles: Article[];
  documents: Document[];
  themes: Theme[];
};

// Query cluser to group queries with similar results from the API
type CacheQueryCluster = {
  // the different queries returning the same results
  queries: string[];
  // the document results : Map<documentPath, {algo: fulltext|prequalified|semantic, cdtnId}>
  results: Map<string, { algo: string; cdtnId: string }>;
};

type Cache = {
  clusters: Map<number, CacheQueryCluster>;
  queryMap: Map<string, number>;
};

export { Document, SearchResult, Cache, CacheQueryCluster };
