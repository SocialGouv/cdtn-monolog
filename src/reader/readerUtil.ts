import { IDataFrame } from "data-forge";
import { setDate, subMonths } from "date-fns";
import { pipe } from "fp-ts/lib/function";
import { getOrElse, none, Option } from "fp-ts/Option";

import { removeArr } from "./themeNames";

export const SERVICE_URL = "https://code.travail.gouv.fr/";

export const urlToPath = (url: string): string => {
  if (url.startsWith(SERVICE_URL)) {
    return url.slice(SERVICE_URL.length);
  } else {
    return url;
  }
};

export const formatDate = (date: Date): string =>
  date.toISOString().split("T")[0];

export const getLastDays = (n: number, ref: Date): string[] => {
  // a day in milliseconds
  const dayMillis = 24 * 60 * 60 * 1000;

  const createDate = (i: number) =>
    new Date(ref.getTime() - (i + 1) * dayMillis);

  return [...Array(n).keys()].map(createDate).map(formatDate);
};

/**
 *  return days of previous month
 *  month : 12 / year : 2020 will return all days in november 2020
 * @param month
 * @param year
 */
export const getDaysInPrevMonth = (month: number, year: number): string[] => {
  const date = subMonths(new Date(year, month, 1, 12), 1); // remove one month --> last finished month
  const currentMonth = date.getMonth();

  const days = [];
  while (date.getMonth() === currentMonth) {
    days.push(new Date(date));
    date.setDate(date.getDate() + 1);
  }
  return days.map(formatDate);
};

export const getLastMonthsComplete = (
  ref: Option<Date> = none,
  n: Option<number> = none
): string[][] => {
  // const end = setDate(getOrElse(() => new Date())(ref), 1);

  const end = pipe(
    ref,
    getOrElse(() => new Date()),
    (d: Date) => setDate(d, 1)
  );

  // we want the 3 last month by default
  const d = 3;
  const nMonths = pipe(
    n,
    getOrElse(() => d),
    (i) => Array(i).keys(),
    (a) => Array.from(a)
  );

  return nMonths
    .map((i) => subMonths(end, i))
    .map((d) => getDaysInPrevMonth(d.getMonth(), d.getFullYear()));
};

export const removeThemesQueries = (dataset: IDataFrame): IDataFrame => {
  return dataset.where((row) => !removeArr.has(row["query"]));
};

export const actionTypes = {
  feedback: "feedback",
  feedback_category: "feedback_category",
  feedback_suggestion: "feedback_suggestion",
  home: "home",
  search: "search",
  searchCC: "cc_search",
  selectCC: "cc_select",
  selectRelated: "select_related",
  selectResult: "select_result",
  selectSuggestion: "select_suggestion",
  themes: "themes",
  visit: "visit_content",
};
