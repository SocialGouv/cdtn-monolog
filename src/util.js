export const SERVICE_URL = "https://code.travail.gouv.fr/";

export const urlToPath = (url) => {
  if (url.startsWith(SERVICE_URL)) {
    return url.slice(SERVICE_URL.length);
  } else {
    return url;
  }
};

export const actionTypes = {
  visitType: "visit_content",
  selectSuggestionType: "select_suggestion",
};
