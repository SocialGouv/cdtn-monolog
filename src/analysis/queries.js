// here we analyze search queries
import { getRouteBySource } from "@socialgouv/cdtn-sources";
import * as murmur from "murmurhash-js";
import PQueue from "p-queue";

import { triggerSearch } from "../cdtnApi";
import * as datasetUtil from "../dataset";
import { logger } from "../logger";
import { actionTypes } from "../util";

const reportType = "query";

// we dedulicate queries and only select the one appearing
// at least {minOccurence} times
const deduplicateQueries = (dataset, minOccurence) =>
  dataset
    .where((a) => a.type == actionTypes.search)
    .groupBy((r) => r.query)
    .select((group) => ({ count: group.count(), query: group.first().query }))
    .inflate()
    .where((r) => r.count >= minOccurence)
    .getSeries("query")
    .toArray()
    .filter((a) => a)
    .map((q) => q.toLowerCase());

// get all search queries and build a cache of the responses
export const buildCache = async (dataset, minOccurence = 0) => {
  const pqueue = new PQueue({ concurrency: 4 });

  // we get all unique queries
  const queries = deduplicateQueries(dataset, minOccurence);

  logger.info(
    `Calling API for ${queries.length} queries, this might take some time`
  );

  const pSearches = Array.from(queries).map((query) =>
    pqueue.add(() =>
      triggerSearch(query)
        .then((json) => ({
          // use base64 to reduce cache size and ease comparison
          // documents: Buffer.from(JSON.stringify(json.documents)).toString(
          //   "base64"
          // ),
          documents: json.documents,
          query,
        }))
        .catch(() => {
          logger.error("Cannot retrieve results for query :" + query);
          return { documents: [], query };
        })
    )
  );

  const results = await Promise.all(pSearches);

  await pqueue.onIdle();

  logger.info("API calls completed.");

  // grouping queries : Map<documentResults, list of queries>
  /** @type{Map.<*, Array.<string>>} */
  const groups = new Map();

  // grouping query and docs : Map<query, documentResults>
  /** @type{Map.<*, Array.<string>>} */
  const resultCache = new Map();

  results.forEach(({ query, documents }) => {
    if (!groups.has(documents)) {
      groups.set(documents, [query]);
    } else {
      groups.get(documents).push(query);
    }

    resultCache.set(query, documents);
  });

  const queryClusters = [...groups.values()];

  // a map to get the query group id from any query : Map<query, clusterIndex>
  /** @type{Map.<string, number>} */
  const queryMap = new Map(
    queryClusters.flatMap((queryCluster, i) =>
      queryCluster.map((query) => [query, i])
    )
  );

  // we build a list of clusters, containing : an id, a list of queries, the API documents
  const cache = new Map(
    queryClusters.map((queryGroup, i) => [
      i,
      {
        queries: new Map(queryGroup.map((q) => [q, 0])),
        results: new Map(
          resultCache
            .get(queryGroup[0])
            // we should use ids here once they're available directly in the URL
            .map(({ algo, source, slug }) => [
              "/" + getRouteBySource(source) + "/" + slug,
              { algo, count: 0 },
            ])
        ),
      },
    ])
  );

  return { cache, queryMap };
};

const computeNDCG = (results) => {
  const dcg = [...results.values()].reduce(
    (acc, val, index) => acc + val.count / Math.log2(index + 1 + 1),
    0
  );

  const idcg = [...results.values()]
    .sort((a, b) => b.count - a.count)
    .reduce((acc, val, i) => acc + val.count / Math.log2(i + 1 + 1), 0);

  const ndcg = idcg && dcg ? dcg / idcg : 0;

  return { dcg, idcg, ndcg };
};

const runEvaluation = (counts) => {
  // count queries and results
  [...counts.values()].forEach((obj) => {
    obj.queriesCount = [...obj.queries.values()].reduce((acc, i) => i + acc, 0);
    obj.selectionsCount = [...obj.results.values()].reduce(
      (acc, i) => i.count + acc,
      0
    );
    Object.assign(obj, computeNDCG(obj.results));
  });

  const clusters = [...counts.values()].map((queryCluster) => {
    const { dcg, idcg, ndcg, queriesCount, selectionsCount } = queryCluster;
    const v = queryCluster.results.values().next().value;
    const type = v && v.algo != "pre-qualified" ? "search" : "pre-qualified";

    const queries = (Array.from(queryCluster.queries) || []).map(
      ([query, count]) => ({
        count,
        query,
      })
    );

    const results = (
      Array.from(queryCluster.results) || []
    ).map(([result, { algo, count }]) => ({ algo, count, result }));

    return {
      dcg,
      idcg,
      ndcg,
      queries,
      queriesCount,
      results,
      selectionsCount,
      type,
    };
  });

  return clusters;
};

// non functionnal implementation here, we increment counts,
// might be worth swapping with immutable implementation at one point
export const analyseVisit = (queryMap, counts) => (v) => {
  const actions = v.where((a) =>
    [actionTypes.search, actionTypes.selectResult].includes(a.type)
  );

  // remove duplicates
  const searches = Array.from(
    new Set(
      actions
        .getSeries("query")
        .toArray()
        .filter((q) => q)
        .map((q) => q.toLowerCase())
    )
  );

  // we increment query count and retrieve results lists
  const results = searches.map((q) => {
    const group = queryMap.get(q);
    const count = counts.get(group);

    if (!count) {
      // logger.error("Cannot find results for query : " + q);
      // logger.error(group);
      return new Map();
    }

    count.queries.set(q, count.queries.get(q) + 1);

    return count.results;
  });

  const resultSelections = actions.where(
    (a) => a.type == actionTypes.selectResult
  );

  // if no selection, not much to do
  if (!resultSelections.count()) return;

  // we unfold the result selection object in two columns
  const unfoldedResultSelections = resultSelections.withSeries({
    resultSelectionAlgo: (df) =>
      df.select((row) =>
        row.resultSelection ? row.resultSelection.algo : undefined
      ),
    resultSelectionUrl: (df) =>
      df.select((row) =>
        row.resultSelection ? row.resultSelection.url : undefined
      ),
  });

  const urlSelected = new Set(
    unfoldedResultSelections
      .getSeries("resultSelectionUrl")
      .toArray()
      .filter((q) => q)
  );

  urlSelected.forEach((url) => {
    const incrementSelection = () => {
      for (const r of results) {
        if (r.has(url)) {
          const obj = r.get(url);
          r.set(url, { algo: obj.algo, count: obj.count + 1 });
          return true;
        }
      }
      return false;
    };

    const found = incrementSelection();

    if (!found) {
      // TODO: what should we do here ?
      logger.error(`Selection not in results : ${url}`);
    }
  });
};

const generateIndexReport = (queryClusters) => {
  const [sumNdcg, sumSelectionCount] = queryClusters
    .map((doc) => [doc.ndcg * doc.selectionsCount, doc.selectionsCount])
    .reduce((a, b) => [a[0] + b[0], a[1] + b[1]], [0, 0]);

  const meanSelectionCount = sumSelectionCount / queryClusters.length;

  // identify problems : above mean selection and poor score
  const problems = queryClusters
    .filter(
      (cluster) =>
        cluster.selectionsCount > meanSelectionCount && cluster.ndcg < 0.8
    )
    .map((cluster) => ({
      count: cluster.selectionsCount,

      ndcg: cluster.ndcg,
      // forcing string
      ...cluster.queries[0],
    }));

  return {
    mean: meanSelectionCount,
    ndcg: sumNdcg / sumSelectionCount,
    problems,
  };
};

export const analyse = async (dataset, reportId = new Date().getTime()) => {
  // get all search queries and build cache for each request using CDTN API
  // counts : a map that store each query group with : queries and occurences / results and clicks
  const { cache: counts, queryMap } = await buildCache(dataset);

  // go through each visit and count queries and selection
  datasetUtil
    .getVisits(dataset.reverse())
    .toArray()
    .filter((v) => v.where((a) => a.type == actionTypes.search).count() > 0)
    .forEach(analyseVisit(queryMap, counts));

  // run evaluation
  const evaluatedQueryClusters = runEvaluation(counts)
    // at least 2 queries
    .filter((d) => d.queriesCount >= 2);

  const hashQueries = (cluster) =>
    murmur.murmur3(JSON.stringify(cluster.queries), 42);

  const queryClusterReports = evaluatedQueryClusters.map((cluster) => ({
    ...cluster,
    queryKey: hashQueries(cluster),
    reportId,
    reportType,
  }));

  const queryClusterIndexReport = {
    ...generateIndexReport(evaluatedQueryClusters),
    reportId,
    reportType: reportType + "-index",
  };

  // console.log(JSON.stringify(queryClusterIndexReport, null, 2));

  // and finally build reports
  return [queryClusterIndexReport, ...queryClusterReports];
};
