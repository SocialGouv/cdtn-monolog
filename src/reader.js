// read data from file system or ES and create Dataframe

import * as dataForge from "data-forge";
import "data-forge-fs";
import { getDocuments } from "./elastic";
import { actionTypes } from "./util";

// get last n days before ref
const getLastDays = (n, ref) => {
  // a day in milliseconds
  const dayMillis = 24 * 60 * 60 * 1000;

  const createDate = (i) => new Date(ref.getTime() - (i + 1) * dayMillis);
  const formatDate = (date) => date.toISOString().split("T")[0];

  return [...Array(n).keys()].map(createDate).map(formatDate);
};

const typesToConsider = Object.values(actionTypes);

export const readFromElastic = async (esClient, n, ref = new Date(), index) => {
  const days = getLastDays(n, ref);

  const query = {
    bool: {
      must: [
        {
          bool: { should: days.map((d) => ({ match: { logfile: d } })) },
        },
        { terms: { type: typesToConsider } },
      ],
    },
  };

  const docs = await getDocuments(esClient, index, query);

  // return a Dataframe containing actions
  return new dataForge.DataFrame({ values: docs, considerAllRows: true });
};

export const readFromFile = async (path) => dataForge.readFile(path).parseCSV();
