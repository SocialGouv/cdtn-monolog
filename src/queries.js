import * as es from "./elastic";
import * as Suggestion from "./analysis/suggestion";
import * as Covisit from "./analysis/covisit";

const Queries = (esClient, index) => {
  return {
    getSuggestions: async () => {
      const suggestionsQuery = {
        term: { reportType: Suggestion.reportType },
      };
      return es.getDocuments(esClient, index, suggestionsQuery);
    },

    getCovisitLinks: async (content) => {
      const covisitQuery = {
        bool: {
          must: [
            {
              term: {
                reportType: Covisit.reportType,
              },
            },
            {
              term: {
                content,
              },
            },
          ],
        },
      };

      //   console.log(JSON.stringify(covisitQuery, null, 2));

      const res = await es.getDocuments(esClient, index, covisitQuery);
      if (res.length == 1) {
        return res[0];
      } else {
        throw new Error(`Covisit query failed for content : ${content}`);
      }
    },
  };
};

export { Queries };
