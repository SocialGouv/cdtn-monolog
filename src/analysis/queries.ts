import { IDataFrame } from "data-forge";
import * as murmur from "murmurhash-js";

import { Cache } from "../cdtn/cdtn.types";
import * as datasetUtil from "../reader/dataset";
import { actionTypes } from "../reader/readerUtil";

type QueryGroup = {
  // the different queries returning the same results and their count
  queries: Map<string, number>;
  // the document results : Map<documentPath, {algo: fulltext|prequalified|semantic, cdtnId, count}>
  results: Map<string, { algo: string; cdtnId: string; count: number }>;
};

type Metrics = {
  ndcg: number;
  idcg: number;
  dcg: number;
  queriesCount: number;
  selectionsCount: number;
};

type QueryCluster = Metrics & {
  queries: {
    count: number;
    query: string;
    suggestion: boolean;
  }[];
  results: {
    algo: string;
    count: number;
    result: string;
  }[];
  selectionsRatio: number;
  type: string;
};

const reportType = "query";

const computeNDCG = (results: Map<string, { count: number }>) => {
  const dcg = [...results.values()].reduce(
    (acc, val, index) => acc + val.count / Math.log2(index + 1 + 1),
    0
  );

  const idcg = [...results.values()]
    .sort((a, b) => b.count - a.count)
    .reduce((acc, val, i) => acc + val.count / Math.log2(i + 1 + 1), 0);

  const ndcg = idcg && dcg ? dcg / idcg : 0;

  return { dcg, idcg, ndcg };
};

const runEvaluation = (
  counts: Map<number, QueryGroup>,
  suggestions: Set<string>
): QueryCluster[] => {
  // count queries and results
  [...counts.values()].forEach((obj) => {
    const queriesCount = [...obj.queries.values()].reduce(
      (acc, i) => i + acc,
      0
    );
    const selectionsCount = [...obj.results.values()].reduce(
      (acc, i) => i.count + acc,
      0
    );
    // TODO we ignore types here
    Object.assign(obj, { queriesCount, selectionsCount });
    Object.assign(obj, computeNDCG(obj.results));
  });

  const clusters = [...counts.values()].map((queryCluster) => {
    // TODO see above
    const {
      dcg,
      idcg,
      ndcg,
      queriesCount,
      selectionsCount,
    } = queryCluster as QueryGroup & Metrics;

    const v = queryCluster.results.values().next().value;
    const type = v && v.algo != "pre-qualified" ? "search" : "pre-qualified";

    const queries = (Array.from(queryCluster.queries) || [])
      .map(([query, count]) => ({
        count,
        query,
        suggestion: suggestions.has(query),
      }))
      .sort((a, b) => b.count - a.count);

    const results = (
      Array.from(queryCluster.results) || []
    ).map(([result, { algo, count }]) => ({ algo, count, result }));

    const selectionsRatio =
      selectionsCount && queriesCount ? selectionsCount / queriesCount : 0;

    return {
      dcg,
      idcg,
      ndcg,
      queries,
      queriesCount,
      results,
      selectionsCount,
      selectionsRatio,
      type,
    };
  });

  return clusters;
};

// non functionnal implementation here, we increment counts,
// might be worth swapping with immutable implementation at one point
const analyseVisit = (
  queryMap: Map<string, number>,
  counts: Map<number, QueryGroup>
) => (v: any) => {
  const actions = v.where((a: any) =>
    [actionTypes.search, actionTypes.selectResult].includes(a.type)
  );

  // remove duplicates
  const searches: string[] = Array.from(
    new Set(
      actions
        .getSeries("query")
        .toArray()
        .filter((q: string) => q)
        .map((q: string) => q.toLowerCase())
    )
  );

  // we increment query count and retrieve results lists
  const results = searches.map((q) => {
    const group = queryMap.get(q);
    const count = group ? counts.get(group) : undefined;

    if (!count) {
      // logger.error("Cannot find results for query : " + q);
      // logger.error(group);
      return new Map();
    }

    count.queries.set(q, count.queries.get(q) || 0 + 1);

    return count.results;
  });

  const resultSelections = actions.where(
    (a: any) => a.type == actionTypes.selectResult
  );

  // if no selection, not much to do
  if (!resultSelections.count()) return;

  // we unfold the result selection object in two columns
  /* done beforehand
    const unfoldedResultSelections = resultSelections.withSeries({
      resultSelectionAlgo: (df) =>
        df.select((row) =>
          row.resultSelection ? row.resultSelection.algo : undefined
        ),
      resultSelectionUrl: (df) =>
        df.select((row) =>
          row.resultSelection ? row.resultSelection.url : undefined
        ),
    });
    */

  const urlSelected = new Set(
    // unfoldedResultSelections
    resultSelections
      .getSeries("resultSelectionUrl")
      .toArray()
      .filter((q: string) => q)
  );

  urlSelected.forEach((url) => {
    const incrementSelection = () => {
      for (const r of results) {
        if (r.has(url)) {
          const obj = r.get(url);
          r.set(url, { algo: obj.algo, count: obj.count + 1 });
          return true;
        }
      }
      return false;
    };

    const found = incrementSelection();

    if (!found) {
      // TODO: what should we do here ?
      // logger.error(`Selection not in results : ${url}`);
    }
  });
};

const sums = (queryClusters: QueryCluster[]) =>
  queryClusters
    .map((doc) => [
      doc.ndcg * doc.selectionsCount,
      doc.selectionsCount,
      doc.queriesCount,
    ])
    .reduce((a, b) => [a[0] + b[0], a[1] + b[1], a[2] + b[2]], [0, 0, 0]);

const generateIndexReport = (queryClusters: QueryCluster[]) => {
  const [sumNdcg, sumSelectionCount, sumQueriesCount] = sums(queryClusters);
  const meanSelectionCount = sumSelectionCount / queryClusters.length;
  const meanQueriesCount = sumQueriesCount / queryClusters.length;

  const pqClusters = queryClusters.filter((q) => q.type != "search");
  const lengthPQ = pqClusters.length;
  const [sumNdcgPQ, sumSelectionCountPQ, sumQueriesCountPQ] = sums(pqClusters);

  const esClusters = queryClusters.filter((q) => q.type == "search");
  const lengthES = esClusters.length;
  const [sumNdcgES, sumSelectionCountES, sumQueriesCountES] = sums(esClusters);

  // identify problems : above mean selection and poor score
  const ndcgThreshold = 0.85;
  const selectionRatioThreshold = 0.6;

  const problems = queryClusters
    .filter(
      (cluster) =>
        cluster.queriesCount > meanQueriesCount &&
        (cluster.ndcg < ndcgThreshold ||
          cluster.selectionsRatio < selectionRatioThreshold)
    )
    .map((cluster) => ({
      ndcg: cluster.ndcg,
      queriesCount: cluster.queriesCount,
      query: cluster.queries[0].query,
      selectionCount: cluster.selectionsCount,
      selectionRatio: cluster.selectionsRatio,
      type: cluster.type,
    }));

  return {
    meanQueryCount: meanQueriesCount,
    meanSelectionCount: meanSelectionCount,
    ndcg: sumNdcg / sumSelectionCount,
    prequalified: {
      clusters: lengthPQ,
      meanQueryCount: sumQueriesCountPQ / lengthPQ,
      meanSelectionCount: sumSelectionCountPQ / lengthPQ,
      ndcg: sumNdcgPQ / sumSelectionCountPQ,
      queryCount: sumQueriesCountPQ,
      selectionCount: sumSelectionCountPQ,
      selectionRatio: sumSelectionCountPQ / sumQueriesCountPQ,
    },
    problems,
    search: {
      clusters: lengthES,
      meanQueryCount: sumQueriesCountES / lengthES,
      meanSelectionCount: sumSelectionCountES / lengthES,
      ndcg: sumNdcgES / sumSelectionCountES,
      queryCount: sumQueriesCountES,
      selectionCount: sumSelectionCountES,
      selectionRatio: sumSelectionCountES / sumQueriesCountES,
    },
    sumQueriesCount,
    sumSelectionCount,
  };
};

export const analyse = (
  dataFrame: IDataFrame,
  queryCache: Cache,
  suggestions: Set<string>,
  reportId = new Date().getTime()
) => {
  // get all search queries and build cache for each request using CDTN API
  // counts : a map that store each query group with : queries and occurences / results and clicks
  // const { cache: counts, queryMap } = await buildCache(dataset);

  const queryMap = queryCache.queryMap;

  // we transform the clusters into a "counting" format to increment counts based on logs
  const counts = new Map<number, QueryGroup>();
  queryCache.clusters.forEach((cacheCluster, index) => {
    const queries = new Map(cacheCluster.queries.map((q) => [q, 0]));
    const results = new Map();
    cacheCluster.results.forEach((obj, url) => {
      results.set(url, { count: 0, ...obj });
    });

    counts.set(index, { queries, results });
  });

  /*
  const queryMap = new Map(
    Array.from(queryCache.clusters.entries()).flatMap(([i, { queries }]) =>
      Array.from(queries.keys()).map((q) => [q, i])
    )
  );
  */

  // go through each visit and count queries and selection
  datasetUtil
    .getVisits(dataFrame.reverse())
    .toArray()
    .filter(
      (v) => v.where((a: any) => a.type == actionTypes.search).count() > 0
    )
    .forEach(analyseVisit(queryMap, counts));

  // run evaluation
  const evaluatedQueryClusters = runEvaluation(counts, suggestions)
    // at least 2 queries
    .filter((d) => d.queriesCount && d.queriesCount >= 2);

  const hashQueries = (cluster: QueryCluster) =>
    murmur.murmur3(JSON.stringify(cluster.queries), 42);

  const queryClusterReports = evaluatedQueryClusters.map((cluster) => ({
    ...cluster,
    queryKey: hashQueries(cluster),
    reportId,
    reportType,
  }));

  const queryClusterIndexReport = {
    ...generateIndexReport(evaluatedQueryClusters),
    reportId,
    reportType: reportType + "-index",
  };

  // console.log(JSON.stringify(queryClusterIndexReport, null, 2));

  // and finally build reports
  return [queryClusterIndexReport, ...queryClusterReports];
};
