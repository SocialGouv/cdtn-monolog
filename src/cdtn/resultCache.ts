import { logger } from "@socialgouv/cdtn-logger";
import { getRouteBySource } from "@socialgouv/cdtn-sources";
import { IDataFrame } from "data-forge";
import * as fs from "fs";
import PQueue from "p-queue";
import * as readline from "readline";

import { esClient, LOG_INDEX } from "../es/elastic";
import { actionTypes } from "../reader/readerUtil";
import { Cache, CacheQueryCluster, Document } from "./cdtn.types";
import { triggerSearch } from "./cdtnApi";

// we deduplicate queries and only select the one appearing
// at least {minOccurence} times
export const deduplicateQueries = (dataset: IDataFrame, minOccurence: number): string[] =>
  dataset
    .where((a) => a.type == actionTypes.search)
    .groupBy((r) => r.query.toLowerCase())
    .select((group) => ({ count: group.count(), query: group.first().query }))
    .inflate()
    .where((r) => r.count >= minOccurence)
    .getSeries("query")
    .toArray()
    .filter((a) => a)
    .map((q) => q.toLowerCase());

const buildQueryMap = (queryClusters: string[][]) =>
  new Map<string, number>(queryClusters.flatMap((queryCluster, i) => queryCluster.map((query: string) => [query, i])));

// get all search queries and build a cache of the responses
export const buildCache = async (): Promise<Cache> => {
  /*
  {
  "size": 0,
  "query": {
    "term": {
      "type": "search"
    }
  },
  "aggs": {
    "grouped_queries": {

      "terms": {
        "field": "query",
        "min_doc_count": 10,
        "size": 20000
      }
    }
  }
}

   */
  const pqueue = new PQueue({ concurrency: 16 });

  // we get all unique queries (deduplicate when happening in same visit + min occurence)
  // const visits = getVisits(dataset);
  // const uniqueSearches = DataFrame.concat(visits.select((visit) => toUniqueSearches(visit)).toArray());

  // const queries: string[] = deduplicateQueries(uniqueSearches, minOccurence);

  logger.info("Getting all queries from ES");
  const result = await esClient.search({
    body: {
      size: 0,
      query: {
        bool: {
          must: [],
          filter: [
            {
              match_phrase: {
                type: "search",
              },
            },
            {
              exists: {
                field: "query",
              },
            },
            {
              range: {
                lastActionTimestamp: {
                  format: "strict_date_optional_time",
                  gte: "2023-05-01T00:00:00.000Z",
                  lte: "2023-06-01T23:59:59.999Z",
                },
              },
            },
          ],
          should: [],
          must_not: [],
        },
      },
      aggs: {
        grouped_queries: {
          terms: {
            field: "query",
            min_doc_count: 2,
            size: 20000,
          },
        },
      },
    },
    index: LOG_INDEX,
    size: 20000,
  });

  const queries: string[] = result.body.aggregations.grouped_queries.buckets.map((bucket: any) => bucket.key);

  logger.info(`Calling API for ${queries.length} queries, this might take some time`);

  // for prequalified results, we want to remove "completed results" and only keep PQ since #2037
  const isolatePrequalified = (documents: Document[]) =>
    documents[0].algo == "pre-qualified" ? documents.filter((d) => d.algo == "pre-qualified") : documents;

  const pSearches = Array.from(queries).map((query) =>
    pqueue.add(() =>
      triggerSearch(query)
        .then(({ documents }) => ({
          // use base64 to reduce cache size and ease comparison
          // documents: Buffer.from(JSON.stringify(json.documents)).toString(
          //   "base64"
          // ),
          documents: isolatePrequalified(documents),
          query,
        }))
        .catch(() => {
          logger.error("Cannot retrieve results for query : " + query);
          return { documents: [] as Document[], query };
        })
    )
  );

  const results = await Promise.all(pSearches);

  await pqueue.onIdle();

  logger.info("API calls completed.");

  // grouping queries : Map<documentResults, list of queries>
  const groups = new Map<string, string[]>();

  // grouping query and docs : Map<query, documentResults>
  const resultCache = new Map<string, Document[]>();

  results.forEach(({ query, documents }) => {
    // const resultKey = murmur.murmur3(JSON.stringify(documents), 42);
    const resultKey = JSON.stringify(documents);
    if (!groups.has(resultKey)) {
      groups.set(resultKey, [query]);
    } else {
      groups.get(resultKey)?.push(query);
    }

    resultCache.set(query, documents);
  });

  const queryClusters = [...groups.values()];

  const queryMap = buildQueryMap(queryClusters);

  // we build a list of clusters, containing : an id, a list of queries, the API documents
  const cache: Map<number, CacheQueryCluster> = new Map(
    queryClusters.map((queryGroup, i) => [
      i,
      {
        queries: queryGroup,
        results: new Map(
          resultCache
            .get(queryGroup[0])
            // we should use ids here once they're available directly in the URL
            ?.map(({ algo, cdtnId, source, slug }: Document) => [
              "/" + getRouteBySource(source) + "/" + slug,
              { algo, cdtnId },
            ])
        ),
      },
    ])
  );

  return { clusters: cache, queryMap };
};

export const persistCache = (cache: Cache, output: string): void => {
  const writer = fs.createWriteStream(output);
  cache.clusters.forEach(({ queries, results }) => {
    writer.write(JSON.stringify({ queries, results: [...results] }));
    writer.write("\n");
  });
  writer.end();
};

export const readCache = async (file: string): Promise<Cache> => {
  const readInterface = readline.createInterface({
    crlfDelay: Infinity,
    input: fs.createReadStream(file),
  });

  const cacheB64 = [];

  for await (const line of readInterface) {
    cacheB64.push(JSON.parse(line));
  }

  const clusters: CacheQueryCluster[] = cacheB64.map(({ queries, results }) => ({
    queries: queries,
    // documents: JSON.parse(new Buffer(documents, "base64")),
    results: new Map(results),
  }));

  // const groups = new Map();

  // cacheB64.forEach(({ query, documents }) => {
  //   if (!groups.has(documents)) {
  //     groups.set(documents, [query]);
  //   } else {
  //     groups.get(documents).push(query);
  //   }
  // });

  // const queryGroups = [...groups.values()];

  const cache = new Map<number, CacheQueryCluster>();
  clusters.forEach((obj, idx) => cache.set(idx, obj));

  const queryMap = buildQueryMap([...clusters.values()].map((e) => e.queries));

  return { clusters: cache, queryMap };
};
