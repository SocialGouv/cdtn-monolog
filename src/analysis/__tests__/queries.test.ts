import { DataFrame } from "data-forge";

import { buildDataFrameFromQueries } from "../../__tests__/util";
import { buildCache } from "../../cdtn/resultCache";
import { actionTypes } from "../../reader/readerUtil";
import { analyse } from "../queries";

describe("Query reports", () => {
  const queries = [
    "contrat de travail",
    "apprentissage",
    "apprentissage",
    "grossesse",
    "grossesse",
    "grossesses",
    "Grossesse",
  ];

  const dataset = buildDataFrameFromQueries(queries);

  it("generate query reports", async () => {
    const cache = await buildCache(dataset, 2);

    const s = "/fiche-service-public/contrat-dapprentissage";

    const selections = new DataFrame([
      { query: "apprentissage", type: actionTypes.search, uvi: 1 },
      {
        resultSelection: s,
        type: actionTypes.selectResult,
        uvi: 1,
      },
    ]);

    const reports = analyse(selections, cache, new Set(["apprentissage"]), 42);

    expect(reports).toMatchSnapshot();
  });
});
