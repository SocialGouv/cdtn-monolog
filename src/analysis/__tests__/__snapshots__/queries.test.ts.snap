// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query reports generate query reports 1`] = `
Array [
  Object {
    "meanQueryCount": NaN,
    "meanSelectionCount": NaN,
    "ndcg": NaN,
    "prequalified": Object {
      "clusters": 0,
      "meanQueryCount": NaN,
      "meanSelectionCount": NaN,
      "ndcg": NaN,
      "queryCount": 0,
      "selectionCount": 0,
      "selectionRatio": NaN,
    },
    "problems": Array [],
    "reportId": 42,
    "reportType": "query-index",
    "search": Object {
      "clusters": 0,
      "meanQueryCount": NaN,
      "meanSelectionCount": NaN,
      "ndcg": NaN,
      "queryCount": 0,
      "selectionCount": 0,
      "selectionRatio": NaN,
    },
    "sumQueriesCount": 0,
    "sumSelectionCount": 0,
  },
]
`;
